kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala

```diff
@@ -266,4 +233,58 @@ object ServiceDiscovery {
       case _: Exception => None
     }
   }
+
+  def getInstance(host: String, port: Int): String = {
+    try {
+      val canonicalHostName = InetAddress.getByName(host).getCanonicalHostName
+      s"$canonicalHostName:$port"
+    } catch {
+      case e: Exception =>
+        throw new KyuubiException(s"Failed to resolve canonical host name for $host", e)
+    }
+  }
+
+  def createZkServiceNode(
+      zkClient: CuratorFramework,
+      namespace: String,
+      instance: String,
+      version: String = KYUUBI_VERSION): PersistentEphemeralNode = {
+    val ns = ZKPaths.makePath(null, namespace)
+    try {
+      zkClient
+        .create()
+        .creatingParentsIfNeeded()
+        .withMode(PERSISTENT)
+        .forPath(ns)
+    } catch {
+      case _: NodeExistsException =>  // do nothing
+      case e: KeeperException =>
+        throw new KyuubiException(s"Failed to create namespace '$ns'", e)
+    }
+    val pathPrefix = ZKPaths.makePath(
+      namespace,
+      s"serviceUri=$instance;version=$version;sequence=")
+    var serviceNode: PersistentEphemeralNode = null
+    try {
+      serviceNode = new PersistentEphemeralNode(
+        zkClient,
+        PersistentEphemeralNode.Mode.EPHEMERAL_SEQUENTIAL,
```

Comment:
should we check whether there existing a zknode with same connecting url?
or just create a new one with different sequence id?

Reply:
It does not matter before.

But now we are developing kyuubi-ctl, so customer could create zknode manually, and might create duplicated zk node.